src/index.ts

```diff
@@ -343,28 +346,56 @@ export class MongoLogManager {
   }
 
   /** Clean up log files older than `retentionDays`. */
-  async cleanupOldLogfiles(): Promise<void> {
+  async cleanupOldLogfiles(maxDurationMs = 5_000): Promise<void> {
     const dir = this._options.directory;
     let dirHandle;
     try {
       dirHandle = await fs.opendir(dir);
     } catch {
       return;
     }
+
+    const deletionStartTimestamp = Date.now();
+    // Delete files older than N days
+    const deletionCutoffTimestamp = deletionStartTimestamp - this._options.retentionDays * 86400 * 1000;
+    // Store the known set of least recent files in a heap in order to be able to
+    // delete all but the most recent N files.
+    const leastRecentFileHeap = new Heap<{
+      fileTimestamp: number, fullPath: string
+    }>((a, b) => a.fileTimestamp - b.fileTimestamp);
+
     for await (const dirent of dirHandle) {
+      // Cap the overall time spent inside this function. Consider situations like
+      // a large number of machines using a shared network-mounted $HOME directory
+      // where lots and lots of log files end up and filesystem operations happen
+      // with network latency.
+      if (Date.now() - deletionStartTimestamp > maxDurationMs) break;
```

Comment:
Honestly, most of the time doing this was spent going back and forth on what to do exactly here. I considered:

- To add a time cap to the cleanup operation (this approach)
- To add a limit to the number of files removed
- To add no limit at all
- To not wait for the result of the cleanup operation here/handle it in a separate Promise chain

Ultimately, I picked this because we don’t want to have applications hanging somewhere indefinitely because they’re busy traversing a large log file directory. I think 5s is a reasonable default threshold (and I don’t intend to override it in mongosh/Compass).